engine/dlib/src/dlib/opaque_handle_container.h

```diff
@@ -0,0 +1,280 @@
+     * free the objects themselves.
+     *
+     * @name ~dmOpaqueHandleContainer
+     */
+    ~dmOpaqueHandleContainer();
+
+    /*# container allocate
+     *
+     * Increases the size of the container.
+     *
+     * @name Allocate
+     * @param count [type:uint32_t] Number of items to increase the container with
+     */
+    bool Allocate(uint32_t count);
+
+    /*# container get
+     *
+     * Get the pointer to the object associated with a handle. The handle will be checked for validity,
+     * i.e the handle version must match the object version and the index of the handle must be in a valid
+     * range of the object list. If the handle is invalid, a NULL pointer will be returned.
+     *
+     * @name Get
+     * @param handle [type:HOpaqueHandle] The object handle
+     * @return pointer [type:T*] Return the object associated with the handle If it is valid, NULL otherwise
+     */
+    T* Get(HOpaqueHandle handle);
+
+    /*# container put
+     *
+     * Adds a reference to an object to the list and returns an opaque handle to that object. If the container
+     * is full, an INVALID_OPAQUE_HANDLE will be returned.
+     *
+     * Note: This function DOES NOT check if the object exists before adding it to the list. You will have
+     *       to maintain this manually by checking that with Get(handle) first.
+     *
+     * @name Put
+     * @param obj [type:T*] The object
+     * @return [type:HOpaqueHandle] Return a handle to the object if successfull, otherwise returns an INVALID_OPAQUE_HANDLE.
+     */
+    HOpaqueHandle Put(T* obj);
+
+    /*# container release
+     *
+     * Removes the references to an object based on a handle, if the handle is valid (i.e points to a valid object in the container).
+     *
+     * @name Release
+     * @param handle [type:HOpaqueHandle] The handle to release
+     */
+    void Release(HOpaqueHandle handle);
+
+    /*# container capacity
+     *
+     * Queries the max objects this container can hold.
+     *
+     * @name Capacity
+     * @return [type:uint32_t] Returns the max object references this container can hold.
+     */
+    uint32_t Capacity();
+
+    /*# container full
+     *
+     * Queries if the container is full. The container is full if every slots in the list is taken by some object.
+     *
+     * @name Full
+     * @return [type:bool] Returns true if all object slots are used, and false if there are available slots.
+     */
+    bool Full();
+
+private:
+    T**       m_Objects;
+    uint16_t* m_ObjectVersions;
+    uint32_t  m_Capacity;
+    uint16_t  m_Version;
+
+    // Internal helper functions
+    inline T* GetByIndex(uint32_t i)
+    {
+        assert(i < m_Capacity);
+        return m_Objects[i];
+    }
+
+    uint32_t GetFirstEmptyIndex()
+    {
+        for (int i = 0; i < m_Capacity; ++i)
+        {
+            if (m_Objects[i] == 0)
+            {
+                return i;
+            }
+        }
+        return INVALID_OPAQUE_HANDLE;
+    }
```

Comment:
Future idea:
In order to improve poerformance when finding the free slot, I think we can encode a linked list inside the `T** m_Objects`.
E.g. if we have a "uintptr_t m_FreeList" which points to the first free index, and where the values in the m_Objects are indices;
```
// popping the first free index
uintptr_t index = m_FreeList;
m_FreeList = (uintptr_t)m_Objects[index];
return index;
```

Reply:
Yes, I thought about adding a freelist as well, I'll make a note about it in the code